### DISCUSSION TITLE
Can parallel building cause race issues?

### DISCUSSION THREAD

**Discussion by vaneyli** [ccache/ccache#243] [Questions & answers] [answered]
Hi guys:
1:    Person A wants to compile  d.cpp   and person B wants to compile d.cpp too.   they share a single .ccache.    I think these may be a race issue without a lock in the code of ccache： the same .manifest and the same .o .

2:    The real world is that I would like to build a huge codebase in one machine and they share a single .ccache.    When the two compiles run together, i think these may be some issues.

for cache hit . I ignore the cwd using export CCACHE_NOHASHDIR=true

**Comment by afbjorklund** [comment, marked as answer]
Yes, ccache lives in a parallell environment. No, there shouldn't be any such issues left.

The stats file should be locked, during update. The cache files are updated in a special order.
Worst case, both compiles will run (in parallell) - even if they both will produce similar output

But _is_ important that all files are updated atomically, etc. File a bug report if you see any issues ?

**Comment by jrosdahl** [comment]
[In addition to @afbjorklund's reply]

> the same .manifest and the same .o 

If two persons compile the same code at the same time, there is a race, yes. But the race only implies that one of them will "win", not that the cache becomes corrupt since the files are "renamed into place", which is atomic on Unix systems. (On Windows, there is a small time window where a result could be inconsistent.)
